File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] result = new int[nums.length];
        int n  = result.length;    
        for(int i=0; i < n; i++){
            result[i] = 1;
        }
        
        int prefix  = 1;
        
        for(int i=0; i < n; i++){
            result[i] *= prefix;
            prefix *= nums[i];
        }
        prefix = 1;
        for(int i=n - 1; i >= 0; i--){
            result[i] *= prefix;
            prefix *= nums[i];
        }
        
        return result;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        lst = [1] * len(nums)
        
        
        products_left = 1
        for i in range(len(nums)):
            lst[i]*= products_left
            products_left*= nums[i]
        
        products_right = 1
        for i in range(len(nums) - 1,-1,-1):
            lst[i] *= products_right
            products_right *= nums[i]
            
        return lst
        
        
        
            